### PR TITLE
Implement ObjectCollector struct to collect objects names present in the provided node message

### PR DESCRIPTION
**Pull request by sanyamsinghal**
Implemented a new ObjectCollector struct
 - for every node/msg in the parse tree, `Collect()` method of this struct is called.
 - `Collect()` method extracts the table and function names(qualified)
 
 I mostly focussed on the object collection from DMLs which we currently supporting in uqc
 
 Added unit tests covering general, complex and corner cases in `unsupported_dml_constructs.go`
 Added multiple generic helper functions in queryparser package like `GetMessageField()`, `GetStringField()` etc...
 
 
Queries ignored:
1. Collected schemas is totally different than source schema list, not containing ""

Queries considered:
1. Collected schemas subset of source schema list
2. Collected schemas contains some from source schema list and some extras

Caveats:
There can be a lot of false positives.
For example: standard sql functions like sum(), count() won't be qualified(pg_catalog) in queries generally.
Making the schema unknown for that object, resulting in query consider

### CODE REVIEW

**Comment by sanyamsinghal** [comment]
https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/4115/

File: yb-voyager/src/queryparser/helpers.go [hunk 1]
```diff
@@ -334,3 +307,46 @@ func getOneofActiveField(msg protoreflect.Message, oneofName string) protoreflec
...
+func GetListField(msg protoreflect.Message, fieldName string) protoreflect.List {
+	field := msg.Descriptor().Fields().ByName(protoreflect.Name(fieldName))
+	if field != nil && msg.Has(field) {
+		return msg.Get(field).List()
+	}
+	return nil
+}
+
+// GetQualifiedName extracts the schema and object name from a list.
+func GetQualifiedName(nameList protoreflect.List) (string, string) {
```
Comment (priyanshi-yb) [inline yb-voyager/src/queryparser/helpers.go:342]: nit: Name is not clear as per the return type, should be something `GetObjectAndSchemaName`

File: yb-voyager/cmd/analyzeSchema.go [hunk 2]
```diff
@@ -1694,12 +1693,13 @@ func checker(sqlInfoArr []sqlInfo, fpath string, objType string) {
...
 	if utils.GetEnvAsBool("REPORT_UNSUPPORTED_PLPGSQL_OBJECTS", true) {
-		checkPlPgSQLStmtsUsingParser(sqlInfoArr, fpath, objType)
+		checkPlPgSQLStmtsUsingParser(sqlInfoArr, fpath, objType, schemaList)
 	}
 	checkStmtsUsingParser(sqlInfoArr, fpath, objType)
 }
 
-func checkPlPgSQLStmtsUsingParser(sqlInfoArr []sqlInfo, fpath string, objType string) {
+func checkPlPgSQLStmtsUsingParser(sqlInfoArr []sqlInfo, fpath string, objType string, schemaList string) {
+	parserIssueDetector := queryissue.NewParserIssueDetector(schemaList)
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/cmd/analyzeSchema.go:1702]: i didn't see its usage anywhere else to converted to local variable.
But i get your point, maybe we can just declare at global level but initialise locally when needed.

File: yb-voyager/src/queryissue/object_collector.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		relName := queryparser.GetStringField(msg, "relname")
+		objectName := relName
+		if schemaName != "" {
+			objectName = fmt.Sprintf("%s.%s", schemaName, relName)
+		}
+		log.Debugf("[RangeVar] fetched schemaname=%s relname=%s objectname=%s field\n", schemaName, relName, objectName)
+		c.addObject(objectName)
+
+	// Extract target table names from DML statements
+	case queryparser.PG_QUERY_INSERTSTMT_NODE, queryparser.PG_QUERY_UPDATESTMT_NODE, queryparser.PG_QUERY_DELETESTMT_NODE:
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/src/queryissue/object_collector.go:67]: SELECT statements are already covered.
In case of SELECTs table name can only be in the FROM clause. (handled switch case 1)
Function names can be target list or from clause. (handled switch case 3)

COPY stmts are not handled right now.
I mostly focussed on the object collection from DMLs which we currently supporting in uqc

File: yb-voyager/src/queryissue/queryissue.go [hunk 4]
```diff
@@ -29,12 +31,15 @@ import (
...
+	SourceSchemaList []string
 	// TODO: Add fields here
 	// e.g. store composite types, etc. for future processing.
 }
 
-func NewParserIssueDetector() *ParserIssueDetector {
-	return &ParserIssueDetector{}
+func NewParserIssueDetector(schemaList string) *ParserIssueDetector {
+	return &ParserIssueDetector{
+		SourceSchemaList: strings.Split(schemaList, "|"),
```
Comment (makalaaneesh) [inline yb-voyager/src/queryissue/queryissue.go:41]: This split logic based on "|" is spread out everywhere. Consider having this logic in cmd package somewhere, and take input directly as a string array

File: yb-voyager/src/queryissue/object_collector.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	"strings"
+
+	"github.com/samber/lo"
+	log "github.com/sirupsen/logrus"
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/queryparser"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// ObjectCollector collects unique schema-qualified object names.
+type ObjectCollector struct {
```
Comment (makalaaneesh) [inline yb-voyager/src/queryissue/object_collector.go:29]: Nothing obviously wrong with the current approach, but how about: 
- Convert `TraverseParseTree` to a struct,
- Collect objects by default while traversing. Let that logic live inside `TraverseParseTree` struct instead of `ObjectCollector` struct. 

Reason being, object collection is core to the traversal, it has nothing to do with detecting issues, so I think it can live inside `queryparser` pkg and be part of the traversal logic directly. (instead of being in the queryissue pkg)
In `queryissue` pkg, your code remains the same, but instead of getting objects, schemas from objectCollector, you get it directly from the traversal struct.

What do you think?

Comment (makalaaneesh) [reply yb-voyager/src/queryissue/object_collector.go:29]: ObjectCollector can remain a separate struct if that makes sense, but let `Traversal` struct expose the methods to get the objects/schemas, and have `queryissue` deal directly with traversal struct, not objectCollector struct.

Comment (sanyamsinghal) [reply yb-voyager/src/queryissue/object_collector.go:29]: hmm let me think about this. Moving the collection logic to queryparser packge makes sense, since these are low level details

File: yb-voyager/cmd/analyzeSchema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go:1702]: Yeah, I am not using this in my refactoring PR.

File: yb-voyager/src/queryissue/object_collector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/src/queryissue/object_collector.go:67]: Why we don't support COPY its also DML, and `COPY (<SELECT stmt>) TO` can have select stmt with functions objects and all?

[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/src/queryissue/object_collector.go:67]: Ok the case is - COPY statements having SELECT inside it.
I think that will be covered, since objectCollector will be executed for each node.

@priyanshi-yb  can you share some examples? I can test that with my current code and also add to unit tests.

[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/src/queryissue/object_collector.go:67]: > Ok the case is - COPY statements having SELECT inside it.

yeah that is the case I could think of right with COPY, not sure if there can be any, examples chatGPT gave - 
```
COPY (SELECT id, xmlagg(xmlparse(document xml_column)) AS combined_xml
      FROM my_table
      GROUP BY id) 
TO '/path/to/file.csv' WITH CSV;

COPY (SELECT ctid, xmin, id, data FROM schema_name.my_table) 
TO '/path/to/file_with_system_columns.csv' WITH CSV;
```

**Comment by sanyamsinghal** [comment]
https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/4129/

File: yb-voyager/src/queryissue/object_collector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/src/queryissue/object_collector.go]: Its working. Added these two cases to unit tests.
```
{
			Sql: `COPY (SELECT id, xmlagg(xmlparse(document xml_column)) AS combined_xml
					FROM my_table
						GROUP BY id)
					TO '/path/to/file.csv' WITH CSV;`,
			ExpectedObjects: []string{"xmlagg", "my_table"},
			ExpectedSchemas: []string{""},
		},
		{
			Sql: `COPY (SELECT ctid, xmin, id, data FROM schema_name.my_table) 
							TO '/path/to/file_with_system_columns.csv' WITH CSV;`,
			ExpectedObjects: []string{"schema_name.my_table"},
			ExpectedSchemas: []string{"schema_name"},
		},
```

File: yb-voyager/src/queryissue/object_collector.go (hunk 5 above)
Comment (sanyamsinghal) [reply yb-voyager/src/queryissue/object_collector.go]: done

File: yb-voyager/src/queryparser/helpers.go [hunk 6]
```diff
@@ -115,49 +119,22 @@ func GetStringValueFromNode(nodeMsg protoreflect.Message) string {
...
-	if svalMsg == nil || !svalMsg.IsValid() {
-		return ""
-	}
-
-	// Ensure svalMsg is of type 'pg_query.String'
-	if svalMsg.Descriptor().FullName() != "pg_query.String" {
+func getStringFromAConstMsg(aConstMsg protoreflect.Message) string {
+	svalMsg := GetMessageField(aConstMsg, "sval")
+	if svalMsg == nil {
 		return ""
```
Comment (makalaaneesh) [inline yb-voyager/src/queryparser/helpers.go:127]: I'm a little apprehensive of returning empty strings. It feelsl like we should also be returning an error to indicate that we could not find the string... but looks like there are a lot of uses here. I will think more about this later and get back.

File: yb-voyager/src/queryparser/object_collector.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		relName := GetStringField(msg, "relname")
+		objectName := lo.Ternary(schemaName != "", schemaName+"."+relName, relName)
+		log.Debugf("[RangeVar] fetched schemaname=%s relname=%s objectname=%s field\n", schemaName, relName, objectName)
+		c.addObject(objectName)
+
+	// Extract target table names from DML statements
+	case PG_QUERY_INSERTSTMT_NODE, PG_QUERY_UPDATESTMT_NODE, PG_QUERY_DELETESTMT_NODE:
+		relationMsg := GetMessageField(msg, "relation")
+		if relationMsg != nil {
+			schemaName := GetStringField(relationMsg, "schemaname")
```
Comment (makalaaneesh) [inline yb-voyager/src/queryparser/object_collector.go:65]: nit: this can be taken out to a separate function `GetObjectQualifiedName` or something like that

File: yb-voyager/src/queryparser/traversal.go [hunk 8]
```diff
@@ -204,3 +210,23 @@ func IsScalarKind(kind protoreflect.Kind) bool {
 		protoreflect.BoolKind, protoreflect.StringKind, protoreflect.BytesKind, protoreflect.EnumKind}
 	return slices.Contains(listOfScalarKinds, kind)
 }
+
+func GetSchemaUsed(query string) ([]string, error) {
```
Comment (makalaaneesh) [inline yb-voyager/src/queryparser/traversal.go:214]: Nice! this is what I had in mind, looks clean!

**Review by makalaaneesh — APPROVED** [review]
LGTM
